pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,626 @@
+	crdInformerFactory := crdinformers.NewSharedInformerFactory(crdClient, informerDefaultResync)
+	namespaceInformer := informerFactory.Core().V1().Namespaces()
+	serviceInformer := informerFactory.Core().V1().Services()
+	networkPolicyInformer := informerFactory.Networking().V1().NetworkPolicies()
+	nodeInformer := informerFactory.Core().V1().Nodes()
+	cnpInformer := crdInformerFactory.Crd().V1alpha1().ClusterNetworkPolicies()
+	anpInformer := crdInformerFactory.Crd().V1alpha1().NetworkPolicies()
+	tierInformer := crdInformerFactory.Crd().V1alpha1().Tiers()
+	cgInformer := crdInformerFactory.Crd().V1alpha3().ClusterGroups()
+	grpInformer := crdInformerFactory.Crd().V1alpha3().Groups()
+	externalNodeInformer := crdInformerFactory.Crd().V1alpha1().ExternalNodes()
+
+	addressGroupStore := store.NewAddressGroupStore()
+	appliedToGroupStore := store.NewAppliedToGroupStore()
+	networkPolicyStore := store.NewNetworkPolicyStore()
+	groupStore := store.NewGroupStore()
+	groupEntityIndex := grouping.NewGroupEntityIndex()
+	labelIdentityIndex := labelidentity.NewLabelIdentityIndex()
+
+	networkPolicyController := networkpolicy.NewNetworkPolicyController(client,
+		crdClient,
+		groupEntityIndex,
+		labelIdentityIndex,
+		namespaceInformer,
+		serviceInformer,
+		networkPolicyInformer,
+		nodeInformer,
+		cnpInformer,
+		anpInformer,
+		tierInformer,
+		cgInformer,
+		grpInformer,
+		addressGroupStore,
+		appliedToGroupStore,
+		networkPolicyStore,
+		groupStore,
+		false)
+
+	controllerQuerier := querier.NewControllerQuerier(networkPolicyController, 10349)
+	externalNodeEnabled := true
+	controllerMonitor := NewControllerMonitor(crdClient, nodeInformer, externalNodeInformer, controllerQuerier, externalNodeEnabled)
+
+	return &fakeController{
+		controllerMonitor,
+		crdClient,
+		client,
+		informerFactory,
+		crdInformerFactory,
+	}
+}
+
+func initController(controller *fakeController, stopCh chan struct{}) {
+	controller.informerFactory.Start(stopCh)
+	controller.crdInformerFactory.Start(stopCh)
+	controller.informerFactory.WaitForCacheSync(stopCh)
+	controller.crdInformerFactory.WaitForCacheSync(stopCh)
+}
+
+func TestSyncExternalNode(t *testing.T) {
+	ctx := context.Background()
+	en := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "TestExternalNode",
+			Namespace: "ns1",
+			Labels:    map[string]string{"en": "vm2"},
+		},
+	}
+	tc := []struct {
+		name                 string
+		key                  string
+		existingExternalNode *v1alpha1.ExternalNode
+		expectedAgentCRD     *v1beta1.AntreaAgentInfo
+		expectedError        string
+	}{
+		{
+			name:          "Invalid key format",
+			key:           "namespace/name/error",
+			expectedError: "unexpected key format: \"namespace/name/error\"",
+		},
+		{
+			name:          "Key does not exist",
+			key:           "default/vm2-e8be5",
+			expectedError: "",
+		},
+		{
+			name:                 "key exists",
```

Comment:
Keep the format consistent if you start it with upper case in other cases' name:
```suggestion
			name:                 "Key exists",
```